### DISCUSSION TITLE
[Closed] Extend ColorAnimationExtensions with generic 'ColorTo' extension

### DISCUSSION THREAD

**Discussion by PieEatingNinjas** [CommunityToolkit/Maui#201] [New Feature Discussions]
Currently the `ColorAnimationExtensions` class has 1 method `BackgroundColorTo` which helps animating the color transition of a `VisualElement`'s `BackgroundColor`.

We could (relatively) easily make this method more generic so we can use it for other properties of type `Color` on `VisualElements` as well. 

For example:
`label.ColorTo(l => l.TextColor, color, rate, duration, easing);`
Where the first parameter identifies which property of type `Color` to update/animate. This parameter is of type `Expression<Func<T, Color>>`

Is this something that would be desired for the Toolkit? I'd think that animating the `TextColor` of a `Label` for example is something devs might want.

This `ColorTo` method doesn't need to replace the existing `BackgroundColorTo` method, as the generic `ColorTo` is a bit more complex to use because of the extra parameter. Plus, the existing `BackgroundColorTo` will be more efficient compared to `ColorTo` where you would need to pass-in the `BackgroundColor` property. So both can live side by side. 
Alternatively,  we could overload the `ColorTo` method, one with the new parameter and one without, where the last one updates the BackgroundColor (== current implementation of `BackgroundColorTo`)

Let's discuss! 🙂

**Comment by TheCodeTraveler** [comment]
I like where this idea is going! 

The implementation is a little tough to infer for new users, though. I.e. If I saw this method in intellisense, would I know how to use it without looking at the documentation?

```cs
Task ColorTo(this VisualElement, Func<T, Color> colorProperty, Color color, uint rate, uint duration, Easing easing)
```

Probably not; I'd certainly be confused by the point of the Func.

## Proposal: Extend `ITextStyle`

What if we extend [`ITextStyle`](https://github.com/dotnet/maui/blob/45ba05e1a5c452966073d9101f21712450a61838/src/Core/src/Core/ITextStyle.cs#L8-L24) instead?

```cs
Task TextColorTo(this ITextStyle, Color color, uint rate = 16, uint duration = 250, Easing? easing = null);
```

This would allow us to use it for any UI control that implements `ITextStyle`, like [`ILabel`](https://github.com/dotnet/maui/blob/main/src/Core/src/Core/ILabel.cs), [`IButton`](https://github.com/dotnet/maui/blob/main/src/Core/src/Core/ITextButton.cs), [`IEntry`](https://github.com/dotnet/maui/blob/main/src/Core/src/Core/IEntry.cs), etc.

```cs
var label = new Label();
await label.TextColorTo(Colors.Red, 16, 1500, Easing.SinIn);

var button = new TextButton();
await button.TextColorTo(Colors.Green, 16, 1500);

var entry = new Entry();
await entry.TextColorTo(Colors.Pink);
```

**Reply by PieEatingNinjas** [reply]
I totally understand that the method signature might not be intuitive.
Your proposal is indeed a lot more clear! I have to admit I'm not yet accustomed to the new MAUI interfaces...😅 It would make a lot more sense to go that direction.

**Reply by TheCodeTraveler** [reply]
Right on! 

When you get a chance, open a new discussion for `ITextStyle`. I'd be happy to champion this feature 💯 

And it's probably not a bad idea to peruse the other interfaces to see if there are any other interfaces that also use `Color` that we can add an extension method for too!

**Reply by PieEatingNinjas** [reply]
At first glance, I found properties of type `Color` on the following interfaces:

-  [IActivityIndicator](https://github.com/dotnet/maui/blob/45ba05e1a5c452966073d9101f21712450a61838/src/Core/src/Core/IActivityIndicator.cs)
 - [IPicker](https://github.com/dotnet/maui/blob/45ba05e1a5c452966073d9101f21712450a61838/src/Core/src/Core/IPicker.cs)
 - [IPlaceholder](https://github.com/dotnet/maui/blob/45ba05e1a5c452966073d9101f21712450a61838/src/Core/src/Core/IPlaceholder.cs)
 - [IProgress](https://github.com/dotnet/maui/blob/45ba05e1a5c452966073d9101f21712450a61838/src/Core/src/Core/IProgress.cs)
 - [ISearchBar](https://github.com/dotnet/maui/blob/45ba05e1a5c452966073d9101f21712450a61838/src/Core/src/Core/ISearchBar.cs)
 - [ISlider](https://github.com/dotnet/maui/blob/45ba05e1a5c452966073d9101f21712450a61838/src/Core/src/Core/ISlider.cs)
 - [ISwitch](https://github.com/dotnet/maui/blob/45ba05e1a5c452966073d9101f21712450a61838/src/Core/src/Core/ISwitch.cs)
 - [ITextStyle](https://github.com/dotnet/maui/blob/45ba05e1a5c452966073d9101f21712450a61838/src/Core/src/Core/ITextStyle.cs)

How do you want to go forward with this? Do I already create a new discussion specific for adding an extension method for `ITextStyle`? Or do you want me to create one discussion where we discuss adding extensions all of the above interfaces?

**Reply by TheCodeTraveler** [reply]
Let's break them up into separate discussions. Otherwise we'd have to implement them all in the same PR, and once we add the implementation, samples and unit tests it'd be a massive PR.

**Comment by CliffAgius** [comment]
Closed as answered.
